purchases/src/main/kotlin/com/revenuecat/purchases/caching/DeviceCache.kt

```diff
@@ -201,12 +227,59 @@ internal class DeviceCache(
         offeringsCachedObject.clearCacheTimestamp()
     }
 
-    @Synchronized fun setOfferingsCacheTimestampToNow() {
+    @Synchronized
+    fun setOfferingsCacheTimestampToNow() {
         offeringsCachedObject.updateCacheTimestamp(Date())
     }
 
     // endregion
 
+    // region subscriber attributes
+
+    @Synchronized
+    fun setAttributes(appUserID: String, attributes: Map<String, SubscriberAttribute>) {
+        val currentlyStoredAttributes = getAllStoredSubscriberAttributes(appUserID)
+        val attributesToBeSet = currentlyStoredAttributes.toMutableMap()
+        attributes.forEach { (key, attribute) ->
+            attributesToBeSet[key] = attribute
+        }
+
+        preferences.edit()
+            .putString(getSubscriberAttributesCacheKey(appUserID), attributesToBeSet.toJSONObject().toString())
+            .apply()
+    }
+
+    @Synchronized
+    fun getAllStoredSubscriberAttributes(appUserID: String): Map<String, SubscriberAttribute> =
+        preferences.getString(getSubscriberAttributesCacheKey(appUserID), null)
+            ?.let { json ->
+                try {
+                    JSONObject(json)
+                } catch (e: JSONException) {
+                    null
+                }
+            }?.buildSubscriberAttributes() ?: emptyMap()
+
+    private fun getSubscriberAttributesCacheKey(appUserID: String) =
+        "$subscriberAttributesCacheKey.$appUserID"
+
+    private fun Map<String, SubscriberAttribute>.toJSONObject() =
+        JSONObject().put("attributes", JSONObject().also {
+            this.forEach { (key, subscriberAttribute) ->
+                it.put(key, subscriberAttribute.toJSONObject())
+            }
+        })
+
+    private fun JSONObject.buildSubscriberAttributes(): Map<String, SubscriberAttribute> {
+        val attributesJSONObject = getJSONObject("attributes")
+        return attributesJSONObject.keys().asSequence().mapNotNull { attributeKey ->
+            val attributeJSONObject = attributesJSONObject[attributeKey]
+            if (attributeJSONObject is JSONObject) attributeKey to SubscriberAttribute(attributeJSONObject)
```

Comment:
never but I thought this was safer than casting in the previous line and crashing. What do you think? Maybe it's better to crash and investigate why it's crashing...

Reply:
I made a question about a similar thing on one of my PRs. I'm not sure what we usually go with for these things. I think it'd be a terrible experience for a batch of apps to crash because they use a buggy version of RevenueCat, but then again it'd be nice to find a way for us to see this kind of error. We don't have a cloud place to log, right?